File: frontend/src/app/core/services/user.service.ts
import { Injectable, signal } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, tap } from 'rxjs';
import {
  User,
  UserDetail,
  CreateUserRequest,
  UpdateUserRequest,
  ResetPasswordRequest
} from '../models/user.model';
import { Todo } from '../models/todo.model';

/**
 * Service for user management operations.
 * Includes both user profile management and admin user management functions.
 */
@Injectable({
  providedIn: 'root'
})
export class UserService {
  private readonly API_URL = '/api/users';

  // Signal for reactive user list state
  users = signal<UserDetail[]>([]);

  constructor(private http: HttpClient) {}

  /**
   * Get all users with detailed information (admin only).
   */
  getAllUsers(): Observable<UserDetail[]> {
    return this.http.get<UserDetail[]>(this.API_URL).pipe(
      tap(users => this.users.set(users))
    );
  }

  /**
   * Get assignable users for todo assignment.
   * Regular users see only themselves, admins see all users.
   * Available to all authenticated users.
   */
  getAssignableUsers(): Observable<User[]> {
    return this.http.get<User[]>(`${this.API_URL}/assignable`);
  }

  /**
   * Get a specific user by ID (admin only).
   */
  getUserById(id: number): Observable<UserDetail> {
    return this.http.get<UserDetail>(`${this.API_URL}/${id}`);
  }

  /**
   * Create a new user (admin only).
   */
  createUser(request: CreateUserRequest): Observable<User> {
    return this.http.post<User>(this.API_URL, request);
  }

  /**
   * Update a user (admin only).
   */
  updateUser(id: number, request: UpdateUserRequest): Observable<User> {
    return this.http.put<User>(`${this.API_URL}/${id}`, request);
  }

  /**
   * Delete a user (admin only).
   */
  deleteUser(id: number): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/${id}`);
  }

  /**
   * Get todos for a specific user (admin only).
   */
  getUserTodos(userId: number): Observable<Todo[]> {
    return this.http.get<Todo[]>(`${this.API_URL}/${userId}/todos`);
  }

  /**
   * Reset a user's password (admin only).
   */
  resetUserPassword(userId: number, newPassword: string): Observable<{message: string}> {
    return this.http.put<{message: string}>(`${this.API_URL}/${userId}/reset-password`, {
      newPassword
    });
  }

  /**
   * Check if a username is available (admin only).
   */
  checkUsernameAvailability(username: string): Observable<{available: boolean}> {
    return this.http.get<{available: boolean}>(`${this.API_URL}/check-username`, {
      params: { username }
    });
  }

  /**
   * Update own password (any authenticated user).
   */
  updatePassword(oldPassword: string, newPassword: string): Observable<{message: string}> {
    return this.http.put<{message: string}>(`${this.API_URL}/password`, {
      oldPassword,
      newPassword
    });
  }

  /**
   * Update own username (any authenticated user).
   */
  updateUsername(newUsername: string): Observable<{message: string; username: string}> {
    return this.http.put<{message: string; username: string}>(`${this.API_URL}/username`, {
      username: newUsername
    });
  }
}

File: backend/src/main/java/com/reptilemanagement/rest/service/UserService.java
package com.reptilemanagement.rest.service;

import com.reptilemanagement.persistence.domain.User;
import com.reptilemanagement.persistence.dto.CreateUserRequest;
import com.reptilemanagement.persistence.dto.UpdateUserRequest;
import com.reptilemanagement.persistence.dto.UserDetailDto;
import com.reptilemanagement.persistence.dto.UserDto;
import com.reptilemanagement.exception.ResourceNotFoundException;
import com.reptilemanagement.persistence.mapper.UserMapper;
import com.reptilemanagement.persistence.repository.UserRepository;
import com.reptilemanagement.security.RoleConstants;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * Service for managing users in the system.
 */
@Service
@RequiredArgsConstructor
public class UserService {

    private final UserRepository userRepository;
    private final UserMapper userMapper;
    private final PasswordEncoder passwordEncoder;

    /**
     * Get all users as basic DTOs.
     */
    public List<UserDto> getAllUsers() {
        return userRepository.findAll().stream()
                .map(userMapper::toDto)
                .collect(Collectors.toList());
    }

    /**
     * Get assignable users for todo assignment.
     * Admin users see all users, regular users see only themselves.
     *
     * @param username The username of the requesting user
     * @return List of assignable users
     */
    public List<UserDto> getAssignableUsers(String username) {
        User currentUser = userRepository.findByUsername(username)
                .orElseThrow(() -> new ResourceNotFoundException("Current user not found"));

        // Check if user is admin
        boolean isAdmin = currentUser.getRoles().contains(RoleConstants.ADMIN);

        if (isAdmin) {
            // Admins can assign to anyone
            return getAllUsers();
        } else {
            // Regular users can only assign to themselves
            return List.of(userMapper.toDto(currentUser));
        }
    }

    /**
     * Get all users with detailed information including todo counts.
     */
    public List<UserDetailDto> getAllUsersDetailed() {
        return userRepository.findAll().stream()
                .map(userMapper::toDetailDto)
                .collect(Collectors.toList());
    }

    /**
     * Get a single user by ID.
     */
    public UserDto getUserById(Long id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User not found with id: " + id));
        return userMapper.toDto(user);
    }

    /**
     * Get detailed user information by ID.
     */
    public UserDetailDto getUserDetailById(Long id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User not found with id: " + id));

        return userMapper.toDetailDto(user);
    }

    /**
     * Create a new user.
     */
    @Transactional
    public UserDto createUser(CreateUserRequest request) {
        // Check if username already exists
        if (userRepository.findByUsername(request.getUsername()).isPresent()) {
            throw new IllegalArgumentException("Username already exists: " + request.getUsername());
        }

        User user = userMapper.toEntity(request);
        user.setPassword(passwordEncoder.encode(request.getPassword()));

        // Set roles - default to USER if not specified
        if (request.getRoles() == null || request.getRoles().isEmpty()) {
            user.setRoles(Set.of(RoleConstants.USER));
        } else {
            user.setRoles(new HashSet<>(request.getRoles()));
        }

        User savedUser = userRepository.save(user);
        return userMapper.toDto(savedUser);
    }

    /**
     * Update an existing user.
     */
    @Transactional
    public UserDto updateUser(Long id, UpdateUserRequest request) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User not found with id: " + id));

        if (request.getEmail() != null) {
            user.setEmail(request.getEmail());
        }

        if (request.getPassword() != null && !request.getPassword().isBlank()) {
            user.setPassword(passwordEncoder.encode(request.getPassword()));
        }

        if (request.getRoles() != null) {
            user.setRoles(new HashSet<>(request.getRoles()));
        }

        User updatedUser = userRepository.save(user);
        return userMapper.toDto(updatedUser);
    }

    /**
     * Delete a user.
     */
    @Transactional
    public void deleteUser(Long id) {
        if (!userRepository.existsById(id)) {
            throw new ResourceNotFoundException("User not found with id: " + id);
        }
        userRepository.deleteById(id);
    }

    /**
     * Reset a user's password (admin function).
     */
    @Transactional
    public void resetUserPassword(Long id, String newPassword) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User not found with id: " + id));

        user.setPassword(passwordEncoder.encode(newPassword));
        userRepository.save(user);
    }

    /**
     * Check if a username is available.
     */
    public boolean isUsernameAvailable(String username) {
        return userRepository.findByUsername(username).isEmpty();
    }
}
